### PR TITLE
Clean up string formatting in a variety of places

### PR DESCRIPTION
**Pull request by stephentoub**
Mostly just replacing string.Format calls with equivalent interpolated string uses.  In some cases, I replaced lots of string creations with a StringBuilder or equivalent.

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs [hunk 1]
```diff
@@ -439,15 +440,33 @@ private void Init(object pca)
...
+
+            int count = ConstructorArguments.Count;
+            for (int i = 0; i < count; i++)
+            {
+                if (!first) vsb.Append(", ");
+                vsb.Append(ConstructorArguments[i].ToString());
+                first = false;
+            }
+
+            int count = NamedArguments.Count;
```
Comment (benaadams) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs:459]:
```suggestion
            count = NamedArguments.Count;
```

File: src/libraries/System.Diagnostics.FileVersionInfo/src/System/Diagnostics/FileVersionInfo.Windows.cs [hunk 2]
```diff
@@ -126,20 +126,18 @@ private static uint GetVarEntry(IntPtr memPtr)
...
-            _productName = GetFileVersionString(memIntPtr, string.Format(CultureInfo.InvariantCulture, template, codepage, "ProductName"));
-            _productVersion = GetFileVersionString(memIntPtr, string.Format(CultureInfo.InvariantCulture, template, codepage, "ProductVersion"));
-            _comments = GetFileVersionString(memIntPtr, string.Format(CultureInfo.InvariantCulture, template, codepage, "Comments"));
-            _legalTrademarks = GetFileVersionString(memIntPtr, string.Format(CultureInfo.InvariantCulture, template, codepage, "LegalTrademarks"));
-            _privateBuild = GetFileVersionString(memIntPtr, string.Format(CultureInfo.InvariantCulture, template, codepage, "PrivateBuild"));
-            _specialBuild = GetFileVersionString(memIntPtr, string.Format(CultureInfo.InvariantCulture, template, codepage, "SpecialBuild"));
+            _companyName = GetFileVersionString(memIntPtr, $"\\\\StringFileInfo\\\\{codepage}\\\\CompanyName");
+            _fileDescription = GetFileVersionString(memIntPtr, $"\\\\StringFileInfo\\\\{codepage}\\\\FileDescription"));
```
Comment (benaadams) [inline src/libraries/System.Diagnostics.FileVersionInfo/src/System/Diagnostics/FileVersionInfo.Windows.cs:130]:
```suggestion
            _fileDescription = GetFileVersionString(memIntPtr, $"\\\\StringFileInfo\\\\{codepage}\\\\FileDescription");
```

File: src/libraries/System.Private.CoreLib/src/System/Reflection/CustomAttributeTypedArgument.cs [hunk 3]
```diff
@@ -43,33 +43,47 @@ internal string ToString(bool typed)
...
+                return $"typeof({((Type)Value!).FullName})";
+
+            if (ArgumentType.IsArray)
             {
                 IList<CustomAttributeTypedArgument> array = (IList<CustomAttributeTypedArgument>)Value!;
-
                 Type elementType = ArgumentType.GetElementType()!;
-                string result = string.Format("new {0}[{1}] {{ ", elementType.IsEnum ? elementType.FullName : elementType.Name, array.Count);
+
+                var result = new ValueStringBuilder(stackalloc char[256]);
```
Comment (benaadams) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/CustomAttributeTypedArgument.cs:65]: Needs a `using` for `ValueStringBuilder`?

File: src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs [hunk 4]
```diff
@@ -726,10 +723,7 @@ private object DebuggerDisplayContent
                 get
                 {
                     var displayTarget = _target as IDebuggerDisplay;
-                    return string.Format("{0} Message={1}, Target=\"{2}\"",
-                        Common.GetNameForDebugger(this),
-                        _messageValue,
-                        displayTarget != null ? displayTarget.Content : _target);
+                    return $"{Common.GetNameForDebugger(this)} Message={_messageValue}, Target=\"{(displayTarget != null ? displayTarget.Content : _target))}\"";
```
Comment (benaadams) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs:726]:
```suggestion
                    return $"{Common.GetNameForDebugger(this)} Message={_messageValue}, Target=\"{(displayTarget != null ? displayTarget.Content : _target)}\"";
```

File: src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs [hunk 5]
```diff
@@ -2411,9 +2399,7 @@ private object DebuggerDisplayContent
                 get
                 {
                     var displaySource = _source as IDebuggerDisplay;
-                    return string.Format("Observers={0}, Block=\"{1}\"",
-                        _observersState.Observers.Count,
-                        displaySource != null ? displaySource.Content : _source);
+                    return $"Observers={_observersState.Observers.Count}, Block=\"{displaySource != null ? displaySource.Content : _source)}\"";
```
Comment (benaadams) [inline src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs:2402]:
```suggestion
                    return $"Observers={_observersState.Observers.Count}, Block=\"{(displaySource != null ? displaySource.Content : _source)}\"";
```

File: src/libraries/System.Private.CoreLib/src/System/Reflection/CustomAttributeTypedArgument.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/CustomAttributeTypedArgument.cs:65]: Its ToString cleans up the state.

**Comment by stephentoub** [comment]
Thanks, @benaadams.

**Pull request merged**